fixes #12714 - Show proxy status in index, create a show page for proxy

shlomizadok opened the pull request:
This PR is part of groups of features which will add more content and visibility to the smart proxy.
In this PR:
- A show page for the proxy, with some data on the smart proxy
- Removal of action buttons from the proxies index and add them to the proxy show page
- Addition of each proxy's status to the proxies index (please note that change of style here affects also the style in the about page

shlomizadok committed:
fixes #12714 - Show proxy status in index, and create a show page for proxy

shlomizadok commented:
[test] - Katello unstable

shlomizadok commented:
[test]

lzap commented on app/assets/javascripts/application.js:
```diff
@@ -11,7 +11,7 @@
 //= require two-pane
 //= require vendor
-//= require about
+//= require statuses
```
Does this mean we were loading the above javascript for each individual page? Can we do better and only load this for about and proxy pages now?

ohadlevy replied:
status of what? just proxy status or hosts etc? (maybe a prefix to the name can help)

ohadlevy replied:
this is not address yet.

shlomizadok replied:
Currently statuses of proxy && compute_resources, maybe later for hosts.

ohadlevy replied:
does it make sense to split them to separate files? I personally find it confusing

lzap commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,62 @@
+        }
+      })
+    });
+});
```
This three methods start looking as repeating, can we extract some bits into helper function? I expect more statuse fields to be added soon.

lzap commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -24,7 +27,6 @@ def edit
   end
 
   def ping
-    @proxy = @smart_proxy
```
Looks like no view uses that in Foreman, but are you sure there is no plugin using that member variable? Had to have a reason.

shlomizadok replied:
This was added in https://github.com/theforeman/foreman/commit/8b737c9c7648b3726dadb3b2e4708fcb43af02a8 
I think that `Parameterizable::ByIdName` overrides it anyhow

shlomizadok replied:
Will restore, anyhow

lzap commented on app/views/smart_proxies/index.html.erb:
```diff
@@ -3,22 +3,24 @@
       <tr>
-        <td class="display-two-pane ellipsis"><%= link_to_if_authorized proxy.name, hash_for_edit_smart_proxy_path(:id => proxy).merge(:auth_object => proxy, :authorizer => authorizer) %></td>
+        <td class="ellipsis">
+          <span class="proxy-status" data-url=<%= ping_smart_proxy_path(proxy) %>><%= image_tag '/assets/spinner.gif' %> <%= _("Connecting..") %></span>
```
Missing quotes around data-url?

shlomizadok commented:
@lzap - Thanks for reviewing. I have fixed all the comments

lzap commented:
Ok, it looks fine, but I think except the overall ping, all the other statuses could be implemented as foreman plugin. We are building something that will be eventually replaced with a generic host and its statuses (proxy is actually a host, we just need to register it). I should have raised my voice during our initial scrum, but it's more clearn when I see the code right here.

This would also open doors for other implementations, I already have a prototype of collectd integration plugin that reads the same values from generic hosts. The plugin could easily add more stuff onto the page.

shlomizadok commented:
![image](https://cloud.githubusercontent.com/assets/433583/11779610/00857890-a265-11e5-82d0-fc592236933d.png)

:smile: 
It could be that you are right, however, We still need this page with the minimal data from the proxy.
Also note, [#12716](http://projects.theforeman.org/issues/12716) which opens the option for other plugins to add their data

lzap commented:
Haven't noticed the ticket, that's exactly what I am looking for. Tests failed.

shlomizadok commented:
@lzap - Tests are failing on un-related Katello <-> Bundler issue

lzap commented on app/views/smart_proxies/index.html.erb:
```diff
@@ -1,24 +1,27 @@
       <tr>
-        <td class="display-two-pane ellipsis"><%= link_to_if_authorized proxy.name, hash_for_edit_smart_proxy_path(:id => proxy).merge(:auth_object => proxy, :authorizer => authorizer) %></td>
+        <td class="ellipsis">
+          <span class="proxy-status" data-url="<%= ping_smart_proxy_path(proxy) %>"><%= image_tag '/assets/spinner.gif' %> <%= _("Connecting..") %></span>
```
Ok I think the spinner is enough, spinner _and_ Connecting text is too much. It also reformats the width of the table when all proxies are done (status icons are just 16px or something like that). In other words, keep only the spinner.

During my initial testing, I don't see the spinner for some reason. Maybe asset cache on development setup?

lzap commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,46 @@
+            <td><%= _('Status') %></td>
+            <td>
+              <span class="proxy-show-status" data-url=<%= ping_smart_proxy_path(@smart_proxy) %>>
+                <%= image_tag '/assets/spinner.gif' %> <%= _("Connecting..") %>
```
Please remove from here as well. All of them. Three in total.

lzap commented on app/views/smart_proxies/index.html.erb:
```diff
@@ -1,24 +1,27 @@
+<table class="table table-bordered table-striped table-fixed">
   <thead>
     <tr>
+      <th class="col-md-1"><%= _('Status') %></th>
```
Can we move status as the very last column in the table? Name is the best 1st column imho, we follow that patterne everywhere.

lzap commented:
The show page looks unfinished.

![image](https://cloud.githubusercontent.com/assets/49752/11807091/da613862-a317-11e5-9645-e7875bcec689.png)

I'd start with "Smart proxy [name]" instead of just simple name.

Then I'd group some buttons together, I think we can put Certificates, Autosign and Import Subnets into Refresh Features button (as a drop-down). Edit and Delete can stay as is.

For what reason have you choosen simple instead of two-pane layout which seems more appropriate to me?

lzap commented:
Additional commen on "Status". I think just "Status" is confusing, status of what? I know this is a "ping" command, so we should show either "Responding" or "Communication status" or similar. We can accept just "Status" in the tables, but not on the show page where we should be more explicit.

shlomizadok commented:
@lzap - thanks for the detailed review.
I have removed the "connecting" and reworded "Status -> Communication status"
I have chosen a full page because, more data should be added via plugins (including core plugins)
So at the end it should look like:
![image](https://cloud.githubusercontent.com/assets/433583/11809772/d6691d3c-a331-11e5-8518-e236869ed05c.png)

this PR lays the ground to adding more data (via plugins for instance, as discussed earlier).

Regarding the buttons, I left them as they are, as it looks more similar to host page

lzap commented:
I still see the "Connecting" text which breaks the table while loading on the About page. Can we do just the spinner instead?

I don't insist on the buttons, I just like grouping more. We use it in discovery for example and Foreman users are used to this for proxies (the list page you replaced).

shlomizadok commented:
@lzap, removed the "connecting" from the about page.
Added buttons group to the proxy's show page

shlomizadok commented:
[test]

lzap commented:
Not related, I will merge once develop is fixed, thanks!

domcleal commented:
> this PR lays the ground to adding more data (via plugins for instance, as discussed earlier).

The layout without any plugins is bad though, it's completely empty as per the screenshot at https://github.com/theforeman/foreman/pull/2971#issuecomment-164739840.  The view in a default installation ought to be reasonable.

> Not related, I will merge once develop is fixed, thanks!

Do you mind if I add a bunch of review comments?  I see a few issues which haven't been addressed.

lzap commented:
Please do @domcleal! I also agree the page is too empty, even if we plan to fill it with subsequent PRs in the incoming weeks.

domcleal commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,62 @@
+  });
+
+  $(".proxy-show-status").each(function(index, item) {
+      var item = $(item);
```
indentation's off here

lzap replied:
Two extra opening spaces.

domcleal commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,62 @@
+  item.attr('title',response.message);
+  if(response.success == true) {
+    item.addClass('text-success')
+    item.text(__(response.message));
```
What's this translating?  It looks like it's a version number.

shlomizadok replied:
Currently it is only used by show version method. But we could later use it for other purposes so thought of keeping the translation?

domcleal replied:
It's probably preferable to just translate it in the server's response if you can, since the locale is already set.  Occasionally there's a reason to send an English string to translate on the client, but there's no point here AFAICT.

domcleal commented on app/views/smart_proxies/index.html.erb:
```diff
@@ -1,24 +1,27 @@
+        <td class="ellipsis"><%= link_to_if_authorized proxy.name, hash_for_smart_proxy_path(:id => proxy).merge(:auth_object => proxy, :authorizer => authorizer) %></td>
         <td class="ellipsis"><%= proxy.url %></td>
         <td class="ellipsis"><%= proxy.features.to_sentence %></td>
-        <td><%= action_buttons(*proxy_actions(proxy, authorizer)) %></td>
```
The index should still have some actions, particularly edit (two-pane) and delete.

domcleal commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,46 @@
+        <tbody>
+          <tr>
+            <td><%= _('Registration date') %></td>
+            <td><%= @smart_proxy.created_at.to_s(:long) %></td>
```
This doesn't appear to be localised, it's rendering in a US style here, check it's using the user's TZ.

Edit: it's also a strange thing to put in the UI, we don't do it anywhere else.

shlomizadok replied:
Will remove

domcleal commented:
1) re https://github.com/theforeman/foreman/pull/2971#discussion-diff-47478883, the advantage of keeping JS in the application.js is to reduce the page load times as Turbolinks will be active.  Moving it into per-page JS files will add another page reload.

2) New top-level JS files must be added to config/environments/production.rb.

domcleal commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,62 @@
+  if(response.status == "OK") {
+    item.attr('title', __('Active'));
+    item.addClass('text-success');
+    item.html('<span class="glyphicon glyphicon-ok-sign" />');
```
These icons are different to the host#index page, it should be the same I think.

shlomizadok replied:
Same icons afaik

shlomizadok commented:
@domcleal, @lzap, thank you both for the review and comments.
I have added some data to the proxy's show page, including "services" tab which show some data on the plugins
I have restored the "edit / delete" buttons for the index page.

shlomizadok commented:
[test] - failure unrelated

lzap commented:
That looks so much better than the previous empty one.

![image](https://cloud.githubusercontent.com/assets/49752/11873982/9dcf77f2-a4dd-11e5-8357-df4f10b9b0c7.png)

Couple of UI points:
- It is not clear what those Host numbers represent. Since there are many ways how to associate a Proxy to a Host, can we do better explanation? As a starter, we could do mouse hover action for Locations
- Total number of hosts (the numbers) should be clickable (takes you to Hosts list with search applied).
- The amout of locations is not limited in the view, maybe to showp "top 10" or something giving user a hint that this is just few?
- Sidenote: what's much more interesting are hosts per Subnet/TFTP, DHCP, DNS, Discovery and then Domain/DNS than Location, which is a lightweight container. I often want to see "hosts with managed DHCP records by this smart proxy".

![image](https://cloud.githubusercontent.com/assets/49752/11873989/a874df58-a4dd-11e5-88f0-ccc1e21f1888.png)
- We should tell the user that these are only the "core" services.
- I'd prefer to list all core services putting N/A for those which are not enabled

domcleal commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,73 @@
+          </div>
+        </div>
+        <div class="row">
+          <h4><%= _('Locations') %></h4>
```
As @lzap hinted at, smart proxies might be associated to hosts via other resources than locations.  Locations are entirely optional and not enabled by default - it could be orgs, or simply via subnets etc.

Edit: at the very least, this would require a check to see if the feature is enabled.

lzap replied:
Exactly, for example in non-Satellite context, this bit will not be much useful.

Idea: If these are widgets, user could decide what to put on, we could offer Locations, Organizations (improved) and things like associations via Subnet/Domain.

lzap commented on app/models/smart_proxy.rb:
```diff
@@ -107,6 +108,44 @@ def version
+    rescue *HTTP_ERRORS => exception
+      raise ::Foreman::WrappedException.new exception, N_("Unable to connect to smart proxy")
+    end
+    Rails.cache.write("plugin_#{plugin_name}_proxy_#{self.id}", result, {:expires_in  => 5.minutes })
```
You write to cache, but never read it?

Also, I think it's much more robust to put caching into the ProxyAPI::\* classes than here. We can simply cache each HTTP response. This could be generic method in ProxyAPI: cached_get instead of get. URL and parameters can be formatted as key: "proxy_cached_get_ #{url}_#{params}" in some way.

Expires in should be a configurable option.

lzap commented on app/models/smart_proxy.rb:
```diff
@@ -107,6 +108,44 @@ def version
+        server = ProxyAPI::TFTP.new(:url => url).bootServer
+        result[:success] = true
+        result[:message] = server
+      end
```
These timeouts in model class do not look good.

If we must use Ruby's Timeout, let's move it again into ProxyAPI and let's make this generic as I expect more plugins to use the same technique. Since there is a cached_get method already, this could become a decorator pattern now - we have cached queries and quieries with timeout. This is pseudocode:

```
server = TimeoutAPI(CachedAPI(ProxyAPI::TFTP.new))
```

In ruby one can do decorators much more cleaner: http://nithinbekal.com/posts/ruby-decorators/

Also we should consider using HTTP Ruby library timeout instead of Ruby Timeout which is dangerous: http://www.mikeperham.com/2015/05/08/timeout-rubys-most-dangerous-api/

lzap commented on app/models/smart_proxy.rb:
```diff
@@ -5,6 +5,7 @@ class SmartProxy < ActiveRecord::Base
   include Taxonomix
   include Parameterizable::ByIdName
   audited :allow_mass_assignment => true
+  CORE_FEATURES = %w(dhcp dns puppet puppet_ca realm tftp)
```
Perhaps configurable option?

lzap commented on app/views/smart_proxies/_tftp.html.erb:
```diff
@@ -0,0 +1,18 @@
+      </div>
+    </div>
+  </div>
+</div>
```
I know it can be little bit early and there is a separate ticket, but in this approach, how plugins are supposed to add new entries here? The Services page on Smart Proxy show screen looks like ideal candidate for our widgeting subsystem! Each plugin can provide it's own widget. Plus we get nice and customizable user interface.

This actually applies also to the 1st tab (Overview).

lzap commented:
Ok I thought I will manage to merge this by end of the week, but I am off for holidays from Monday, therefore I hereby resign from the official reviewer. :-) I mean, this is a great discussion to have. Let's not rush code in!

Improvements in this area are great, I hope this to be useful in all contexts and ready to be expanded by either other parts of the core, or plugins. I do agree that monster PRs are not good, but this one is still in good size and can grow in size a bit.

tstrachota commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,73 @@
+    <div class="tab-pane" id="services">
+      <div class="col-md-6">
+        <% @smart_proxy.core_features.each do |feature| %>
+          <%= render :partial => feature, :locals => {:feature => feature} if lookup_context.template_exists?("smart_proxies/#{feature}", [], true) %>
```
I would be good to have the feature partials in a subdirectory so that they don't mix with the main templates.

tstrachota commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,73 @@
+    </div>
+    <div class="tab-pane" id="services">
+      <div class="col-md-6">
+        <% @smart_proxy.core_features.each do |feature| %>
```
Do we actually need to do intersection with the core features? Won't it limit plugins in adding their own services? Checking for existence of the partials seems enough to me (but I might be missing something).

dLobatog commented:
I can pick up the review, I'll set it to WoC because there are open comments & needs to be rebased after Rails 4 has been merged.

shlomizadok commented:
Rebased.
Added cached_get method to the proxy
Removed `core_features`  method as it is not needed.
Moved plugin partials to `smart_proxies/plugins` directory

shlomizadok commented:
[test]

ohadlevy commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,77 @@
+      url:  url,
+      data: {plugin: item.data('plugin')},
+      success: function(response) {
+        console.log(response)
```
do you always want to log it? why inconsistent with the others?

ohadlevy commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,77 @@
+      url:  url,
+      success: function(response) {
+        setItemMessage(item, response);
+      }
```
is there a value in handing error? what happens if there is no response? (ditto for the other ajax below)

shlomizadok replied:
The connection is with Foreman, and we assume it is there. If there is a connection problem with the proxy it is handled on the foreman server side

ohadlevy commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,77 @@
+  } else {
+    item.attr('title', response.message);
+    item.addClass('text-danger')
+    item.html('<span class="glyphicon glyphicon-exclamation-sign" />');
```
consider extracting to a helper function

ohadlevy commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,77 @@
+
+function setItemMessage(item, response) {
+  item.attr('title',response.message);
+  if(response.success == true) {
```
nitpick: do you need to == true?

ohadlevy commented on app/assets/javascripts/two-pane.js:
```diff
@@ -28,6 +28,13 @@ $(document).on('click', "a[href$='new'].new_two_pane", function(e) {
   }
 });
 
+$(document).on('click', "a[href$='edit'].edit_two_pane", function(e) {
```
is this related ? this can effect many pages? (just wondering if it should be its own pr)

shlomizadok replied:
I have checked - this is the only page with css class `edit_two_pane`. I can couple it more closely if you'd - or add it as a separate pr

ohadlevy replied:
other plugins might use it, afair @orrabin had something like that in discovery?

orrabin replied:
I used the class `display-two-pane` and I added it to every place I found that used two-pane.

ohadlevy commented on app/assets/stylesheets/application.scss:
```diff
@@ -541,3 +541,8 @@ span.glyphicon.host-status {
   background: transparent image-url('switch.png') no-repeat 270px 80px;
 }
+
+.big {
```
nitpick: big sounds wrong to me as a css class name:)

ohadlevy replied:
this is unused at the end? on purpose?

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -1,11 +1,16 @@
 class SmartProxiesController < ApplicationController
   include Foreman::Controller::AutoCompleteSearch
-  before_filter :find_resource, :only => [:edit, :update, :refresh, :ping, :version, :destroy]
+  helper :hosts, :reports
```
not sure if its needed, afair all helpers were avail on all controllers?

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -49,6 +54,25 @@ def version
+    rescue Foreman::Exception => exception
+      render :json => {:success => false, :message => exception.message} and return
+    end
+    render :json => {:success => true, :message => tftp_server[:message]}
```
nitpick: it doesn't feel DRY, are you assuming the method would be different in the future? (or maybe its common infrastructure for every plugin author to add a section)?
you could do something like (not tested)

``` ruby

def plugin_version
  plugin = params[:plugin]
  fetch_proxy do 
    response = @smart_proxy.plugin_version(plugin)
    render :json => response
  end
end

private

def fetch_proxy &block
  response = yield
  return {:success => true, :message => response[:message]}
rescue Foreman::Exception => exception
  return {:success => false, :message => exception.message}
end
```

shlomizadok replied:
Done, besides proxy_version which searches on a hash (but refactored it too)

ohadlevy replied:
did you push? i still see the old code

shlomizadok replied:
@ohadlevy - it is the new code :) - take another look

ohadlevy replied:
hmm..  i must be blind, i still see DRY all over :)

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -71,7 +95,7 @@ def action_permission
       when 'refresh'
         :edit
-      when 'ping', 'version'
+      when 'ping', 'version', 'plugin_version', 'tftp_server'
```
if this is going to be an infra for plugins to add, consider setting it up in a method or something

ohadlevy commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -1,19 +1,61 @@
+      end +
+      content_tag(:div, :class => 'col-md-8') do
+        content_tag(:span, {:class => 'plugin-version', :data => {:url => plugin_version_smart_proxy_path(proxy), :plugin => feature}}) do
+          image_tag('/assets/spinner.gif')
```
no helper / css class for this?

shlomizadok commented:
Some screenshots:
Proxies index page:
![proxies_index](https://cloud.githubusercontent.com/assets/433583/11918078/fc5add80-a72a-11e5-9510-aa85f062d082.png)

Proxy show (connected):
![proxy_show_connected](https://cloud.githubusercontent.com/assets/433583/11918081/09d2d198-a72b-11e5-8c22-b54cdeab443a.png)

Proxy show services (connected):
![with_working_services](https://cloud.githubusercontent.com/assets/433583/11918083/1b54b1de-a72b-11e5-8bfa-a703d7aae908.png)

Proxy show (disconnected):
![proxy_show_error](https://cloud.githubusercontent.com/assets/433583/11918086/278bb042-a72b-11e5-8aaa-da7288df4cce.png)

Proxy show services (one service, disconnected):

![proxy_services_error](https://cloud.githubusercontent.com/assets/433583/11918087/340e0fae-a72b-11e5-9e4e-83f889fe8371.png)

ohadlevy commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -1,19 +1,61 @@
-     if SETTINGS[:unattended] and proxy.has_feature?('DHCP')
-       display_link_if_authorized(_("Import subnets"), hash_for_import_subnets_path(:smart_proxy_id => proxy))
-     end,
+  def render_plugin_version(proxy, feature)
```
why a helper vs in view?

shlomizadok replied:
Used more than once, so I won't repeat

ohadlevy replied:
yes, but can't this be a partial?

ohadlevy commented on app/models/smart_proxy.rb:
```diff
@@ -91,11 +91,35 @@ def has_feature?(feature)
   def version
     result = {}
     begin
-      Timeout::timeout(20) do
```
why dropping the timeout? is it because its ajax ?

shlomizadok replied:
Seems that timeout is handled at the [ProxyApi::Resource](https://github.com/theforeman/foreman/blob/develop/lib/proxy_api/resource.rb#L8) and it is using Net::HTTP timeout which seems to be safer

ohadlevy commented on app/models/smart_proxy.rb:
```diff
@@ -91,11 +91,35 @@ def has_feature?(feature)
+    result
+  end
+
+  def plugin_version(plugin_name)
```
similar method from above example could be reused here DRY

ohadlevy commented on app/views/about/index.html.erb:
```diff
@@ -81,7 +81,7 @@
                     <td><%= link_to compute.name, compute %></td>
                     <td><%= compute.provider_friendly_name %></td>
-                    <td><div class="compute-status" data-url=<%= ping_compute_resource_path(compute) %>><%= _("Connecting..") %></div></td>
+                    <td><div class="compute-status" data-url=<%= ping_compute_resource_path(compute) %>><%= image_tag '/assets/spinner.gif' %></div></td>
```
we have css class called .spinner-placeholder

ohadlevy replied:
missed this spinner instance

ohadlevy commented on config/routes.rb:
```diff
@@ -190,11 +190,13 @@
       post 'ping'
       put 'refresh'
       get 'version'
+      post 'plugin_version'
```
why POST ?

shlomizadok replied:
Because I am sending data (the plugin name)

ohadlevy replied:
but you are expecting to GET the data right? a more proper HTTP usage might be:
`GET plugin_version?name=...`

ohadlevy commented on lib/proxy_api/resource.rb:
```diff
@@ -84,5 +84,11 @@ def put(payload, path = "")
+
+    def cached_get(path = nil, payload = {})
+      Rails.cache.fetch("proxies/#{url}", :expires_in => 3.minutes) do
+        get(path, payload)
```
do we store somewhere when was the data last refreshed? (we could use that to display in the UI)

ohadlevy replied:
I would assume we should cache on the controller level.. does clicking of refresh features invalidate the cache?

lzap replied:
I still think there are benefits in having this here. It's a separate method name so this is safe from misusage.

The current implementation will need to have separate caching code per particular function.

ohadlevy commented:
@shlomizadok in the last screenshot with a puppet error, how can see the actual error (it just say error in red - i have no idea what is it)

shlomizadok commented:
@ohadlevy, thanks for the review. in the puppet error, when hovering with a mouse it show the error

shlomizadok commented:
Hope that now it is not :exclamation::repeat:
@dLobatog / @ohadlevy - will you take another look please?

ohadlevy commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,84 @@
+    var item = $(item);
+    var url = item.data('url');
+    $.ajax({
+      type: 'post',
```
why is this POST?

ohadlevy commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,76 @@
+        <div class="row">
+          <div class="col-md-4">
+            <%= _('Features') %>
+            <%= refresh_proxy_icon(@smart_proxy, authorizer) %>
```
this redirects to index page, i would assume refresh from here should leave me on the same page?

shlomizadok replied:
It returns to the proxy show page for me

tbrisker replied:
If an error occurs it redirects to the index page. To me it makes sense to have this refresh run as ajax instead of reloading the entire page.

ohadlevy replied:
:+1:

dLobatog replied:
It works for me - in any case if the current refresh already reloads the page I don't see the problem with this one doing the same, particularly when there's so much content to change if features change. I'd move it to be next to the Feature list (col-md-8 next to it) though, it looks a bit weird on the 'labels' column.

shlomizadok commented:
@ohadlevy fixed your comments

ohadlevy commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,84 @@
+  });
+
+  $(".plugin-version").each(function(index, item) {
+    var item = $(item);
```
atm there are multiple queries to the proxy at the same time, does it make sense to wait until the initial ping query is done successfully and only then do follow up queries?

ohadlevy commented on app/assets/javascripts/statuses.js:
```diff
@@ -0,0 +1,84 @@
+      url:  url,
+      data: {plugin: item.data('plugin')},
+      success: function(response) {
+        if (response)
```
when would response be false?

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -41,12 +46,21 @@ def refresh
+  def plugin_version
+    plugin_name = params[:plugin]
+    render :json => {:success => false, :message => _('Plugin name cannot be blank')} and return if plugin_name.blank?
+    Rails.cache.fetch("proxy_#{@smart_proxy.id}/#{plugin_name}/version", :expires_in => 3.minutes) do
```
does it make sense to cache the value per proxy version? AFAIR you get it all in one hash, so maybe cache the entire hash?

ohadlevy commented on app/models/smart_proxy.rb:
```diff
@@ -89,21 +89,36 @@ def has_feature?(feature)
   def version
-    result = {}
+    fetch_proxy_data do
+      ProxyAPI::Version.new(:url => url).version
```
nitpick: you are creating `ProxyAPI::Version.new(:url => url)` multiple times, consider extracting it:

``` ruby

def version_api
  @version_proxy ||= ProxyAPI::Version.new(:url => url)
end
```

ohadlevy commented on app/views/about/index.html.erb:
```diff
@@ -32,8 +32,8 @@
                   <td><%=h proxy.features.to_sentence %></td>
-                  <td><div class="proxy-status" data-url=<%= ping_smart_proxy_path(proxy) %>><%= _("Connecting..") %></div></td>
-                  <td><div class="proxy-version" data-url=<%= version_smart_proxy_path(proxy) %>><%= _("Connecting..") %></div></td>
+                  <td><div class="proxy-status" data-url=<%= ping_smart_proxy_path(proxy) %>><%= image_tag '/assets/spinner.gif' %></div></td>
```
1. why do you hardcode /assets? is that a new requirement?
2. why not use the css class with the spinner instead?

ohadlevy commented on lib/proxy_api/version.rb:
```diff
@@ -11,6 +11,10 @@ def proxy_versions
     end
 
+    def plugin_versions
+      proxy_versions["modules"]
```
since you are calling proxy_versions multiple times, cache it, e.g.

``` ruby
    def proxy_versions
      @version_response ||= parse(get)
    rescue => e
      raise ProxyException.new(url, e, N_("Unable to detect version"))
    end
```

ohadlevy commented:
@shlomizadok also, when creating a new smart proxy - I think it should redirect to the show page?

ohadlevy commented:
@shlomizadok also, if no locations are attached to the proxy, it does not show anything, imho it should say that no locations are attached to this proxy. further, you need to test this with context (e.g. in a location, vs in another location, vs any context).

shlomizadok commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,65 @@
+          </div>
+          <div class="col-md-8">
+            <span class="proxy-show-status" data-url="<%= ping_smart_proxy_path(@smart_proxy) %>">
+              <%= spinner%><
```
Who added that `<` ? :imp: 
(will remove when tests end)

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -67,11 +70,25 @@ def destroy
+    return true if Rails.env.test?
+    # As memcached does not support delete_matched, we need to delete each
+    Rails.cache.delete("proxy_#{@smart_proxy.id}/versions")
+    Rails.cache.delete("proxy_#{@smart_proxy.id}/tftp_server")
```
should this be delegated to the proxy_status class?

ohadlevy commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def proxy_versions
+    Rails.cache.fetch("proxy_#{@proxy_id}/versions", :expires_in => 3.minutes) do
```
nitpick: I would extract the expiry to a variable / method

lzap replied:
Or global setting.

ohadlevy commented on app/views/smart_proxies/index.html.erb:
```diff
@@ -1,3 +1,4 @@
+<% javascript('statuses') %>
```
forgot to the change the js name

ohadlevy commented on app/views/smart_proxies/plugins/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,17 @@
+    <div class="row">
+      <h4><%= _('Puppet CA Details') %></h4>
+    </div>
+    <%= render :partial => 'smart_proxies/plugins/plugin_version', :locals => {:feature => feature, :proxy => @smart_proxy } %>
```
should we show the puppet certificates here?

tbrisker replied:
I think the certificates should be a tab added when the PuppetCA feature is present

ohadlevy commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,65 @@
+          </div>
+          <div class="col-md-8">
+            <span class="proxy-version" data-url="<%= ping_smart_proxy_path(@smart_proxy) %>">
+              <%= spinner %><
```
ditto

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -1,11 +1,15 @@
   include Foreman::Controller::AutoCompleteSearch
-  before_filter :find_resource, :only => [:edit, :update, :refresh, :ping, :version, :destroy]
+
+  before_filter :find_resource, :only => [:show, :edit, :update, :refresh, :ping, :version, :plugin_version, :tftp_server, :destroy]
```
need to remove version and plugin_version

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -67,11 +64,25 @@ def destroy
       when 'refresh'
         :edit
-      when 'ping', 'version'
+      when 'ping', 'plugin_version', 'tftp_server'
```
need to remove plugin_version

ohadlevy commented on app/views/about/index.html.erb:
```diff
@@ -29,11 +29,11 @@
               <tbody>
               <% @smart_proxies.each do |proxy| %>
-                <tr>
+                <tr class="proxy-show"  data-url=<%= ping_smart_proxy_path(proxy) %>>
```
nitpick: extra whitespace and extra ">"

ohadlevy replied:
do we need both the proxy show class here and for each <tr>?

shlomizadok replied:
yup. javascript calls on each proxy-show

ohadlevy commented on app/views/smart_proxies/plugins/_plugin_version.html.erb:
```diff
@@ -0,0 +1,8 @@
+<div class="row">
+  <div class="col-md-4"><%= _('Plugin version') %></div>
+  <div class="col-md-8">
+    <span class="plugin-version" data-url="<%= plugin_version_smart_proxy_path(proxy) %>" data-plugin="<%= feature %>">
```
data-url not required?

ohadlevy commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,66 @@
+      <div class="col-md-6">
+        <% @smart_proxy.features.pluck('LOWER(name)').uniq.each do |feature| %>
+          <% feature_erb = "smart_proxies/plugins/#{feature}" %>
+          <%= render :partial => feature_erb, :locals => {:feature => feature} if lookup_context.template_exists?(feature_erb, [], true) %>
```
what if no templates are avail? it would be an empty tab?

shlomizadok replied:
Either empty, or give notice that no features are enabled. Which is better?

ohadlevy commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -0,0 +1,103 @@
+}
+
+function setPluginVersion(item, response) {
+  "use strict";
```
nitpick: DRY all of the functions are very similar.

shlomizadok commented:
fixed comments and added support for plugins w/o template (see openscap & discovery for instance)
![image](https://cloud.githubusercontent.com/assets/433583/11976686/7aaee8c0-a983-11e5-81bc-d899c848e085.png)

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -67,11 +64,25 @@ def destroy
+  def requested_data(data_name)
+    data = @smart_proxy.public_send(data_name)
+    render :json => {:success => true, :message => data}
+  rescue Foreman::Exception => exception
```
i think we need to catch more exceptions here (at somewhere else) I just had Errno::ENOENT  exception (No such file or directory @ rb_sysopen when opening ssl certificate)

shlomizadok replied:
I think it should be added to `HTTP_ERRORS` as it is supposed to catch all connection issues and raise one `Foreman::Exception`. Easier to read and parse ("One error to rule them all")
![image](https://cloud.githubusercontent.com/assets/433583/11976877/4560e5fe-a985-11e5-9495-79db88f4a6e2.png)

shlomizadok replied:
Added to `HTTP_ERRORS`

ohadlevy commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -0,0 +1,103 @@
+}
+
+function setProxyVersion(item, response) {
+  item.attr('title',response.message.version);
```
why set the version both here and in the text? I find it confusing to see the same value on hover.

ohadlevy commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -0,0 +1,103 @@
+  var self = this;
+
+  this.getVersions = function() {
+    $.ajax({
```
nitpick: you can switch to getJSON

shlomizadok commented:
[test] is not triggered?

tbrisker commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -0,0 +1,100 @@
+      success: function (response) {
+        populateData(response, self.item);
+      }.bind(this),
+      error: function (response) {
```
Since the handling is the same for success and error, you can simply use `complete`.

shlomizadok replied:
Changed. Thanks :+1:

tbrisker replied:
no you didn't :wink:

shlomizadok replied:
Yeah, I reverted cause it didnt work

tbrisker commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -23,16 +27,10 @@ def edit
     old_features = @smart_proxy.features
     if @smart_proxy.refresh.blank? && @smart_proxy.save
+      delete_cached_versions
       msg = @smart_proxy.features == old_features ? _("No changes found when refreshing features from %s.") : _("Successfully refreshed features from %s.")
```
The first message shows even if there are new or removed features.

tbrisker replied:
to fix this, change line 31 and here to use `@smart_proxy.features.to_a` which will force a query instead of lazy relation that is evaluated on comparison (and obviously gives the same result)

tbrisker replied:
This seems to actually be broken in develop already...

ohadlevy commented:
[test] oom errors

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -493,6 +493,10 @@ def new_child_fields_template(form_builder, association, options = { })
   end
 
+  def spinner
+    '<span class="spinner-placeholder" />'.html_safe
```
from some reason, the image does not show up for me in a span

ohadlevy replied:
ok, the size of the span is 0 as there is no text, hence its not visible.
consider adding something like:

``` css
 background: image-url('spinner.gif') no-repeat top left;
 display: inline-block;
```

ohadlevy replied:
also note that the position of the status icon and spinner is not the same.

ohadlevy replied:
that seems to be because of  `margin: 20% 30% 0 40%;` in the spinner-placeholder css.

ohadlevy commented:
@shlomizadok new assets are missing in config/environments/production.rb.

ohadlevy commented:
why does the version of the proxy is in green?

![proxy-version-color](https://cloud.githubusercontent.com/assets/20167/12009767/ce386854-ac8f-11e5-87c3-76960a4b6b49.png)

ohadlevy commented:
what does tftp false means?

![tftp-false](https://cloud.githubusercontent.com/assets/20167/12009770/f8fa5192-ac8f-11e5-9cd2-298e05d2f97a.png)

ohadlevy commented on app/views/smart_proxies/plugins/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,37 @@
+    </div>
+    <div class="row">
+      <div class="col-md-4">
+        <%= _('Pending certificates') %>
```
this looks nice, maybe we should highligh in a different color that there are pending certs?

ohadlevy commented on app/views/smart_proxies/plugins/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,37 @@
+        <% pending = SmartProxies::PuppetCA.find_by_state(@smart_proxy, "pending").count rescue _('N/A')%>
+        <%= display_link_if_authorized(pending, hash_for_smart_proxy_puppetca_index_path(:smart_proxy_id => @smart_proxy, :state => 'pending').
+                                              merge(:auth_object => @smart_proxy, :permission => 'view_smart_proxies_puppetca', :authorizer => authorizer)) %>
+      </div>
```
missing autosign counters?

ohadlevy commented on app/views/smart_proxies/plugins/_realm.html.erb:
```diff
@@ -0,0 +1,16 @@
+<div class="row">
+  <div class="col-md-12">
+    <div class="row">
+      <h4><%= _('Realm Details') %></h4>
```
I think we can drop the word "details", when having multiple features the context is clear imho:
![details](https://cloud.githubusercontent.com/assets/20167/12009785/9ca36c2a-ac90-11e5-9733-dea39ffc14c4.png)

ohadlevy commented:
@shlomizadok did we discuss showing the proxy used orgs/locations?

ohadlevy commented:
while waiting for api calls, the UI moves around: e.g.
![spinner](https://cloud.githubusercontent.com/assets/20167/12009790/0c7eff64-ac91-11e5-8fe7-6d7ad07c67d6.png)

and right after it completes:

![complete](https://cloud.githubusercontent.com/assets/20167/12009793/1f56349a-ac91-11e5-8c0a-32dcc112801e.png)

ohadlevy commented on app/views/about/index.html.erb:
```diff
@@ -29,11 +29,11 @@
                   <td><%=h proxy.features.to_sentence %></td>
-                  <td><div class="proxy-status" data-url=<%= ping_smart_proxy_path(proxy) %>><%= _("Connecting..") %></div></td>
-                  <td><div class="proxy-version" data-url=<%= version_smart_proxy_path(proxy) %>><%= _("Connecting..") %></div></td>
+                  <td><div class="proxy-show-status"><%= spinner %><</div></td>
```
extra "<"

lzap commented:
I haven't tested myself, but I don't see Locations in the code while it is on the screenshots. I think this is the right decition is to take it off the patch for now allowing us to properly discuss it. :clap:

shlomizadok commented:
Latest adds puppet in a different tab:
![image](https://cloud.githubusercontent.com/assets/433583/12010472/0c99fe52-acb1-11e5-9bf5-268cd92d23e6.png)
And the comments above

ohadlevy commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,87 @@
+          <% if lookup_context.template_exists?(feature_erb, [], true) %>
+            <%= render :partial => feature_erb, :locals => {:feature => feature} %>
+          <% else %>
+            <%= render :partial => "smart_proxies/plugins/no_template", :locals => {:feature => feature} unless feature.start_with?('puppet') %>
```
should we have a services tab if there are no services?
also, what if the only services are puppet*? i have an empty services tab.

ohadlevy commented on app/views/smart_proxies/plugins/puppet/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,47 @@
+      </div>
+      <div class="col-md-8">
+        <% autosign = ProxyAPI::Puppetca.new(:url => @smart_proxy.url).autosign.count %>
+        <%= display_link_if_authorized(autosign, hash_for_smart_proxy_autosign_index_path(:smart_proxy_id => @smart_proxy).
```
since this entire tab does not load over ajax, if there is error here the entire show page wont render.

ohadlevy commented on app/views/smart_proxies/plugins/puppet/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,47 @@
+        <%= _('Pending certificates') %>
+      </div>
+      <div class="col-md-8">
+        <% pending = SmartProxies::PuppetCA.find_by_state(@smart_proxy, "pending").count %>
```
I would also show revoked certificate number
and expiry date of the CA.

ohadlevy commented on app/views/smart_proxies/plugins/puppet/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,43 @@
+          <%= _('%s certificates') % state.capitalize %>
+        </div>
+        <div class="col-md-8">
+          <% state_count =  SmartProxies::PuppetCA.find_by_state(@smart_proxy, state).count rescue _('Error getting data for %s certificates') % state %>
```
this will "eat" the exception. IMHO you need to change it to a service class, that can later on be used in the api too.

ohadlevy commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,93 @@
+<% title(_('Smart Proxy: %s') % @smart_proxy.to_label) %>
+<%= smart_proxy_title_actions(@smart_proxy, authorizer) %>
+<% puppet_features_lowered = %w(puppet puppet_ca) %>
+<% service_features = @smart_proxy.features_to_lowercase - puppet_features_lowered %>
```
imho there is too much logic in the view layer :( and too little tests of all of the scenarios.

shlomizadok commented:
Latest removed proxy tab (will be added as a separate PR [issue #12720](http://projects.theforeman.org/issues/12720))

dLobatog commented on app/views/smart_proxies/plugins/_plugin_version.html.erb:
```diff
@@ -0,0 +1,8 @@
+<div class="row">
+  <div class="col-md-4"><%= _('Plugin version') %></div>
```
I don't think this should show up if the feature doesn't come from a plugin. Just `version` is fine or even better, nothing when the feature is part of the default features in the proxy, and 'plugin version' when it comes from a plugin.

shlomizadok replied:
Renamed to `Version`. I think it's a nice to have data, even for core plugins

dLobatog commented on app/views/about/index.html.erb:
```diff
@@ -29,11 +29,11 @@
-                  <td><div class="proxy-status" data-url=<%= ping_smart_proxy_path(proxy) %>><%= _("Connecting..") %></div></td>
-                  <td><div class="proxy-version" data-url=<%= version_smart_proxy_path(proxy) %>><%= _("Connecting..") %></div></td>
+                  <td><div class="proxy-show-status"><%= spinner %></div></td>
+                  <td><div class="proxy-version"><%= spinner %><</div></td>
```
It looks like there's some mistake here? When the spinner is still loading, a open tag character shows up next to it. 
![screenshot from 2015-12-28 12-44-50](https://cloud.githubusercontent.com/assets/598891/12018268/ef103b3e-ad60-11e5-841a-6bc4696a111a.png)

shlomizadok replied:
oops, forgot to remove... removing :)

shlomizadok replied:
Removed

dLobatog commented on app/views/about/index.html.erb:
```diff
@@ -29,11 +29,11 @@
               <% @smart_proxies.each do |proxy| %>
-                <tr>
+                <tr class="proxy-show"  data-url="<%= ping_smart_proxy_path(proxy) %>">
                   <td><%= link_to_if_authorized proxy.name, hash_for_edit_smart_proxy_path(proxy) %></td>
```
Now that there's a `show` page, I think it'd be more appropriate to link to the smart_proxy page.

shlomizadok replied:
:+1: replaced

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def delete_cached_versions
+    return true if Rails.env.test?
```
Why?

dLobatog replied:
?

shlomizadok replied:
Wasn't this removed?

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,46 @@
+  private
+
+  def fetch_proxy_data &block
+    begin
```
Unless you're passing &block to another method, you can do without the argument (it's faster), it's included by default in all methods http://mudge.name/2011/01/26/passing-blocks-in-ruby-without-block.html

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -67,11 +63,18 @@ def destroy
   private
 
+  def requested_data(data_name)
+    data = @smart_proxy.public_send(data_name)
```
why do we need to go via the proxy at all? can't we just go directly to the status class(es)?

dLobatog commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -1,19 +1,54 @@
-        display_link_if_authorized(_("Autosign"), hash_for_smart_proxy_autosign_index_path(:smart_proxy_id => proxy).
-                                merge(:auth_object => proxy, :permission => 'view_smart_proxies_autosign', :authorizer => authorizer))]
-     end,
+    [ display_link_if_authorized(_("Edit"), hash_for_edit_smart_proxy_path(:id => proxy), :class => 'edit_two_pane'),
```
Why removing all of the previous options? I saw them under 'Select action' on `smart_proxies/:id` show, but now something that used to take me 1 click now is 2 clicks away.

shlomizadok replied:
Reverted. Re-organized order though (edit, puppet ca actions, delete)

dLobatog commented on app/models/smart_proxy.rb:
```diff
@@ -88,22 +86,25 @@ def has_feature?(feature)
   end
 
+  def features_to_lowercase
+    features.pluck("LOWER(REPLACE(name, ' ', '_'))").uniq
```
Why is this needed? Same for the method that calls it in the helper.

shlomizadok replied:
It is needed to load the erb files.
The feature is `DHCP`, the erb file is `_dhcp.html.erb`.
It comes more handy in places like `Puppet CA`, where the erb file should be `_puppet_ca.html.erb`

dLobatog commented on app/models/smart_proxy.rb:
```diff
@@ -88,22 +86,25 @@ def has_feature?(feature)
-      raise ::Foreman::WrappedException.new exception, N_("Unable to connect to smart proxy")
-    end
-    result
+    proxy_status.proxy_versions
```
`delegate :version, :tftp_server, :to => :proxy_status`, handle the 'no tftp feature' there directly.

shlomizadok replied:
usted tiene razón
![image](https://cloud.githubusercontent.com/assets/433583/12019018/b809063a-ad73-11e5-8a32-6ad75391195d.png)

ohadlevy commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,52 @@
+                       Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError]
+
+  def initialize(proxy, opts = {})
+    @proxy = proxy
```
how about?
`raise ::Foreman::Exception.new(N_('Must specify a Smart Proxy to use')) unless proxy.is_a?(SmartProxy)`

shlomizadok replied:
It's you code... will change :wink:

shlomizadok replied:
Added

dLobatog commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -1,19 +1,54 @@
-       display_link_if_authorized(_("Import subnets"), hash_for_import_subnets_path(:smart_proxy_id => proxy))
-     end,
+  def generate_links_for(proxy_sub_model)
+    return _("None Found") if proxy_sub_model.empty?
```
nitpick, `None found`

dLobatog commented on app/models/smart_proxy.rb:
```diff
@@ -88,22 +88,16 @@ def has_feature?(feature)
-    end
-    result
+  def features_to_lowercase
+    features.pluck("LOWER(REPLACE(name, ' ', '_'))").uniq
```
If this is strictly to load partials & it has everything to do with the presentation layer, shouldn't this be in the helper?

shlomizadok replied:
Yeah, will remove

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,55 @@
+                       Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError]
+
+  def initialize(proxy, opts = {})
+    raise ::Foreman::Exception.new(N_('Must specify a Smart Proxy to use')) unless proxy.is_a?(SmartProxy)
```
Kinda confused at this check? I don't think it's worth protecting anything here, but in any case if an object walks like a SmartProxy, swims like a SmartProxy and quacks like a SmartProxy, I [call that a SmartProxy](https://en.wikipedia.org/wiki/Duck_typing).  

What I'm trying to say is that even though I doubt anyone will call ProxyStatus.new with anything other than a SmartProxy, if a plugin has an object that behaves like a SmartProxy it should probably be allowed as well?

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,55 @@
+  attr_reader :proxy, :cache_duration, :api
+  alias_method :version, :proxy_versions
+
+  def fetch_proxy_data &block
```
&block isn't necessary here as an argument and makes it slower, `yield` is enough

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,55 @@
+class ProxyStatus
```
:clap: - 1 minor nitpick, this looks like a model, not a service, shouldn't it be on /models?

ohadlevy replied:
imho no, it does not own data, rather has a specific function, for example see https://www.netguru.co/blog/service-objects-in-rails-will-help

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,55 @@
+
+  # TODO: extract to another status implementation
+  def tftp_server
+    raise ::Foreman::Exception.new(N_('No TFTP feature for %s') % proxy.to_label) unless proxy.has_feature?('TFTP')
```
I'd say it'd be safe to implement a 'UnavailableFeatureProxyStatus` for all features that are not available yet. Basically implementing a [NullObject pattern](https://robots.thoughtbot.com/rails-refactoring-example-introduce-null-object), & making it return something neutral

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,55 @@
+    Rails.cache.delete("proxy_#{proxy.id}/tftp_server")
+  end
+
+  def cache_key
```
`versions_cache_key`? Just so that the name is more descriptive

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,55 @@
+
+  private
+
+  attr_reader :proxy, :cache_duration, :api
```
Looks like :api is not used/called from anywhere? Moreover it can be retrieved calling `proxy_version_api`

ohadlevy replied:
right, its a leftover from another class we experimented with

shlomizadok replied:
Actually re-used it. Please see latest commit

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -0,0 +1,55 @@
+    @cache_duration = opts[:cache_duration] || 3.minutes
+  end
+
+  def proxy_version_api
```
No need to prefix methods with `proxy` within a ProxyStatus IMO, it makes it redundant e.g `proxy_status.proxy_version_api` vs `proxy_status.api_version` - ditto below on `versions`

dLobatog commented on test/unit/smart_proxy_test.rb:
```diff
@@ -59,11 +59,15 @@ class SmartProxyTest < ActiveSupport::TestCase
     test "should succeed" do
       proxy = smart_proxies(:one)
-      fake_version = {:version => '1.11'}
+      fake_version = {'version' => '1.11'}
```
You can make it expect the cache to contain the version if you don't want to clear the cache for this test.

dLobatog commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -1,19 +1,60 @@
-     if SETTINGS[:unattended] and proxy.has_feature?('DHCP')
-       display_link_if_authorized(_("Import subnets"), hash_for_import_subnets_path(:smart_proxy_id => proxy))
-     end,
+  def generate_links_for(proxy_sub_model)
```
Move to application_helper if it's this general. It looks like it doesn't have to do much with smart_proxy and it would work on other models.

dLobatog commented:
Generally :+1: , pretty impressed with how it looks already. I commented a few issues but nothing too major. I'm only missing more testing, especially around ProxyStatus and the controller changes.

Thanks @shlomizadok !

dLobatog commented on app/views/smart_proxies/plugins/_dns.html.erb:
```diff
@@ -0,0 +1,16 @@
+<div class="row">
+  <div class="col-md-12">
```
This should be the default on bootstrap but I can see why you're doing it, otherwise the content is misaligned. I think this html needs to be slightly modified to look like the Overview page. Here's what I did:

``` html
<div class="row">
  <h3><%= _('DNS') %></h3>
</div>
<%= render :partial => 'smart_proxies/plugins/plugin_version', :locals => {:feature => feature, :proxy => @smart_proxy } %>
<div class="row">
  <div class="col-md-4">
    <%= _('Domains') %>
  </div>
  <div class="col-md-8">
    <%= generate_links_for(@smart_proxy.domains).html_safe %>
  </div>
</div>
```

That's the same layout as in the Overview page, now it looks a little different.
![screenshot from 2015-12-28 16-29-29](https://cloud.githubusercontent.com/assets/598891/12020814/816e2a62-ad80-11e5-990a-0fc138003156.png)

dLobatog commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,75 @@
+    <div class="tab-pane active in" id="properties">
+      <div class="col-md-6">
+        <div class="row">
+          <h4><%= _('Details') %></h4>
```
I'd say h3, like on Host#new puppetclasses tab

shlomizadok commented:
@dLobatog - Added tests for tftp_version action. Please note that testing ping & version are the same
Also changed html according to your recommendation

ohadlevy commented on app/helpers/application_helper.rb:
```diff
@@ -395,6 +395,15 @@ def documentation_button(section = "")
+  def generate_links_for(sub_model)
+    return _("None found") if sub_model.empty?
+    links = []
+    sub_model.each do |model|
```
sub_model.map.. ?

shlomizadok replied:
Changed.

shlomizadok commented:
Added tests for ProxyStatus class and UI looks like:
Index page
![image](https://cloud.githubusercontent.com/assets/433583/12037343/c2c48ebc-ae56-11e5-87b2-5271b0e9bdf1.png)

Show page#overview
![image](https://cloud.githubusercontent.com/assets/433583/12037359/e956e41c-ae56-11e5-8389-42dda6e634ed.png)

Show page#services
![image](https://cloud.githubusercontent.com/assets/433583/12037387/1da4218a-ae57-11e5-8b6d-493661cbec3e.png)

Show page#overview (in error mode)
![image](https://cloud.githubusercontent.com/assets/433583/12037405/4e1cbe6c-ae57-11e5-8786-bfa8953e151b.png)

Show page#services (in error mode)
![image](https://cloud.githubusercontent.com/assets/433583/12037415/6cc27870-ae57-11e5-812f-2f6d9fa807a4.png)

dLobatog commented on app/views/smart_proxies/plugins/_no_template.html.erb:
```diff
@@ -0,0 +1,7 @@
+<div class="row">
+  <div class="col-md-12">
```
col-md-12?

shlomizadok replied:
removed

dLobatog commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -0,0 +1,74 @@
+            <% if lookup_context.template_exists?(feature_erb, [], true) %>
+              <%= render :partial => feature_erb, :locals => {:feature => feature} %>
+            <% else %>
+              <%= render :partial => "smart_proxies/plugins/no_template", :locals => {:feature => feature} unless feature.start_with?('puppet') %>
```
This seems redundant if `service_features` is set to `proxy_features_besides_puppet` already.

shlomizadok replied:
removed

dLobatog commented:
@shlomizadok Thanks for the update! I'd say it's basically ready. I went through it thoroughly and this is all I found. All minor issues, so IMO :+1: to merge after it's fixed.
- [ ] The whole exception is not displayed on index page when I hover over the status icon. However, on the show page, it's displayed properly next to Communication status.
  ![screenshot from 2015-12-29 20-39-37](https://cloud.githubusercontent.com/assets/598891/12041005/de72911a-ae6d-11e5-8399-e40133fad9e3.png)
- [ ] Because the `no_template` partial still uses the old HTML, it's misaligned.
  ![screenshot from 2015-12-29 20-49-49](https://cloud.githubusercontent.com/assets/598891/12040999/d5136e82-ae6d-11e5-84e3-0ebd12e5f52c.png)
- [ ] Instead of the numerous red 'Error' messages, which display no useful information to the user, I think it'd be better to substitute them by the same error icon I see on communication status. That provides a bit more info about what happened.
- [ ] What happened to the BMC partial?
- [ ] Tabs should resize according to content, I substituted the min-height on `proxy-content.tab-content` for 0px to do that. I also removed the margin-top property you set on `.proxy-content.tab-content .row`. It conflicts with line-height inadvertently and makes text look worse than the default I think. I'd even argue we can do without the whole `proxy-content` class and it looks nice, but it's your call.
  ![screenshot from 2015-12-29 21-46-32](https://cloud.githubusercontent.com/assets/598891/12041875/b76e19f6-ae75-11e5-8a68-52d81a559774.png)

shlomizadok commented:
- [x] tooltip on index
  ![image](https://cloud.githubusercontent.com/assets/433583/12042591/0a31ad10-ae85-11e5-9456-a68135e9c384.png)
- [x] `no_template` aligned
  ![image](https://cloud.githubusercontent.com/assets/433583/12042607/39ad38fc-ae85-11e5-93bd-8baf5b68898e.png)
- [x] exclamation sign + info instead of errors
  ![image](https://cloud.githubusercontent.com/assets/433583/12042620/6189f95a-ae85-11e5-9ea5-0828a44cb180.png)
- [ ] Will add at a later phase?
- [x] `min-height: 0`, .removed `proxy-content.tab-content .row`
  ![image](https://cloud.githubusercontent.com/assets/433583/12042659/cd05f986-ae85-11e5-8112-1aef1764fb39.png)

dLobatog commented:
Merged as 7cdaddab0289b13528a041fb2432147dfa8de47a, thanks @shlomizadok!

dLobatog commented:
& @ohadlevy , @lzap , @tbrisker, @tstrachota

dscaplan commented:
I want to ensure that we are addressing the need to identify the association of hosts to capsules. This is a mandate for Capsule (Smart Proxy) observability. This can be accomplished with a host page under capsules and / or a scoped search for the host table with the capsule name used as a matcher. 99% of downstream customers place all Services on a single capsule.
